ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkUtils.kt

```diff
@@ -4,19 +4,35 @@
 
 package io.ktor.client.engine.okhttp
 
+import io.ktor.client.features.*
 import io.ktor.http.*
 import kotlinx.coroutines.*
 import okhttp3.*
 import okhttp3.Headers
 import java.io.*
+import java.net.*
 import kotlin.coroutines.*
 
 internal suspend fun OkHttpClient.execute(request: Request): Response = suspendCancellableCoroutine {
     val call = newCall(request)
     val callback = object : Callback {
 
         override fun onFailure(call: Call, cause: IOException) {
-            if (!call.isCanceled) it.resumeWithException(cause)
+            if (call.isCanceled) {
+                return
+            }
+
+            val mappedException = when (cause) {
+                is ConnectException -> HttpConnectTimeoutException()
+                is SocketTimeoutException -> if (cause.message?.contains("connect") == true) {
+                    HttpConnectTimeoutException()
+                } else {
+                    HttpSocketTimeoutException()
+                }
+                else -> cause
+            }
+
+            it.resumeWithException(mappedException)
```

Comment:
Same problem with missing cancellation wrapping and potential connection refused errors

Reply:
FIxed I guess.